### ISSUE TITLE
twisted.mail.smtp.sendmail should use ESMTP

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7257]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7257|
|Type|defect|
|Created|2014-04-27 07:18:10Z|
|Branch|https://github.com/twisted/twisted/tree/esmtp-sendmail-7257-3|

Currently, `twisted.mail.smtp.sendmail` uses `SMTPSenderFactory`, which only speaks plain SMTP (and therefore will never use STARTTLS), and connects over TCP, therefore providing no TLS at all.

Since most major email providers are switching to TLS-only (Google Apps/GMail, Microsoft Office 365), use `ESMTPSenderFactory` instead, with `requireTransportSecurity` set to `False`, so that it doesn't break existing applications, but will upgrade using STARTTLS if it can, which is much better, and will work with the extra functionality these services now provide.

<details><summary>Searchable metadata</summary>

```
trac-id__7257 7257
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__ 
branch__branches_esmtp_sendmail_7257_3 branches/esmtp-sendmail-7257-3
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1398583090000000 1398583090000000
changetime__1429025191196692 1429025191196692
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

(In [42564]) Branching to esmtp-sendmail-7257.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

This patch enables ESMTP in sendmail.

By passing extra kwargs to sendmail, you can require encryption, or use authentication. If the other server does not speak ESMTP, it will degrade to SMTP gracefully. If it does speak ESMTP and STARTTLS is available, it will upgrade the connection to an encrypted one.

Other things done in this branch are fixing a minor bug in ESMTP (where authenticationRequired could be bypassed if the server was SMTP-only) and fixing up some nearby code to be a bit more in line with the standard. There's also a fix to make ESMTP support cancelling deferreds properly.

This includes tests for making sure sendmail honours the ESMTP options, and for ESMTP downgrading being cancelled if TLS or Authentication is required.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @hawkowl|
|-|-|

Thanks.

  1. The ticket description only talks about switching from SMTP to ESMTP - not about these various other bug fixes or feature enhancements.  Please limit the scope of individual tickets to one improvement or bug fix.  Sometimes it does make sense to combine multiple improvements but usually only when their implementations are deeply intertwined and therefore difficult to separate.  That doesn't appear to be the case here.  I suggest splitting out the feature enhancements that aren't just "use ESMTP if possible" and the bug fixes.
  1. `ESMTPClient.heloFallback` should have a docstring instead of a comment.
  1. The line wrapping for the string literal in `esmtpAUTHMalformedChallenge` does not follow the conventions used elsewhere in Twisted.  Please use () to split string literals across multiple lines.
  1. That method and many of the other similar methods which are modified here are missing docstrings.
  1. The `@type` documentation for the `auth` parameter accepted by `registerAuthenticator` is ... weird.  It should say `L{IClientAuthentication} implementor`.
  1. The `requireAuthentication` parameter now accepted by `sendmail` is mis-documented.
  1. The `username` and `password` parameters accepted by that function are missing documentation as well.  Also... I'm not sure this is the correct interface to present for ESMTP authentication.  For example, how will this cope with a server that offers GSSAPI?
  1. Throughout, byte string literals should use the `b"..."` syntax.
  1. `test_requireTLSAndHELOFallbackSucceedsIfOverTLS` only makes the transport provide `ISSLTransport` *after* the connection is made.  This isn't very realistic.
  1. `test_honorsESMTPArguments` and `test_sendmailDefaults`:
    1. have a lot of calls to `assertEqual`
    1. have no docstrings
  1. No build has been forced, please do that for the next review pass.
  1. For the newly introduced features, there should be some documentation apart from the raw API documentation.  An example or a howto or something would be good.  I suspect there are even some existing examples that you will be able to modify and simplify to demonstrate these features.


Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

I'll spin the relevant bugfixes into [[#7258](https://github.com/twisted/twisted/issues/7258)](https://github.com/twisted/twisted/issues/7258) soon, and file tickets for any other things once I pull out those changes. I'll then remake this ticket into just the feature, and just have #7258 be a requirement for this to land.

Thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

(In [43837]) Branching to esmtp-sendmail-7257-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

I've merged this stuff forward, and improved test coverage for sendmail. This patch is just those tests and making sendmail use ESMTP.

As far as whether username/password is the right way to expose it, it's how it's exposed in ESMTPSenderFactory, and how smtplib does it, so it seems okay for this high-level interface.

I'd like to get the code bits reviewed, and then I'll do some usage documentation once I know that this is what I need to document. :)

Builds spun, all green so far <3

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="derwolfe's avatar" src="https://avatars.githubusercontent.com/u/641357?s=50" width="50" height="50">](https://github.com/derwolfe)<a name="note_10"></a>|@derwolfe set owner to @hawkowl|
|-|-|

Hi Hawkowl,

Thanks for you work on this!

t.m.smtp.sendmail
        - on line 2100, the epytext markup should read "This interface is intended to be a replacement for L{smtplib.SMTP.sendmail}"
        - @type declarations for each of the parameters would be helpful.
        - the docstring would be a bit more readable if there were a newline between each of the @param sections.

t.m.test.test_smtp
        - line 1836, the docstring for test_honorsESMTPArguments seems a bit off; I believe you could omit the word passes and it would say what you're wanting.
        - Would it be helpful to have a test showing that if a username and/or password is not provided, SMTP authentication doesn't occur?

I also agree that only expecting a username and password for authentication is acceptable, as this is how the standard lib's sendmail and t.m.smtp.ESMTPSenderFactory work.

Please resubmit for review once you've made the changes and added how-to/usage documentation.

Thanks again for your work!

Best,
Chris / herrwolfe

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

(In [44364]) Branching to estmp-sendmail-7257-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl commented|
|-|-|

(In [44365]) Branching to esmtp-sendmail-7257-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl removed owner|
|-|-|

Hi Chris, thanks for the review.

As for the comments:

- Fixed.
- Done.
- Done.
- Revamped the docstrings.
- I think that's outside of the scope of this ticket -- that's implemented in the underlying ESMTPClientFactory.

I've written up some documentation, and I'm fairly happy with it, so I'm putting this up for review again.

-hawkie

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

(Also spun the builders to make sure they're happy)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Hi,

Have you pushed your changes?

I can not see them on branch esmtp-sendmail-7257-3

https://github.com/twisted/twisted/compare/trunk...esmtp-sendmail-7257-3

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl removed owner|
|-|-|

Sigh, I pushed it into the wrong (typo'd) branch.

Pushed up to the correct branch, respun the builders.

**Comment by twisted-trac** [comment]
|<img alt="txmanish's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|txmanish set owner to @hawkowl|
|-|-|

Hi hawkowl,

Thanks a lot for your continued changes :). Just 3 nitpick changes:

1. Could you use `t.i.task.react` in all the `sendmail_*.py` example files? This makes the code simpler since you don't need to stop the reactor manually and this should be the preferred way of writing one-off scripts.
2. Please use  `self.assertIs` instead of `self.assertIdentical` in `test_messageFilePassthrough` test.
3. It will be nice to change `hasattr(msg, 'read')` to `getattr(msg, 'read', None)` and check if returned `read` method is `callable` but that is outside the scope of this ticket.

Everything else looks good to me :)



Everything else looks

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_18"></a>|@hawkowl commented|
|-|-|

Thanks for the review!

I've fixed [#11338](https://github.com/twisted/twisted/issues/11338) and #2. Waiting for the buildbots to spin, then will merge when theyre green.

- Hawkie

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_19"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44412]) Merge esmtp-sendmail-7257-3: Sendmail should use ESMTP

Author: hawkowl
Reviewers: exarkun, herrwolfe, txmanish
Fixes: [#7257](https://github.com/twisted/twisted/issues/7257)
